src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,197 @@
+            Node<T> x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> l = last;
+        Node<T> newNode = new Node<>(l, value, null);
+        last = newNode;
+        if (l == null) {
+            first = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+
+        if (index == size) {
+            add(value);
+            return;
+        }
+
+        Node<T> current = getElementByIndex(index);
+        Node<T> prev = current.prev;
+        Node<T> newNode = new Node<>(prev, value, current);
+        current.prev = newNode;
+
+        if (prev == null) {
+            first = newNode;
+        } else {
+            prev.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexWithEquals(index);
+
+        if (index < (size >> 1)) {
+            Node<T> x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x.value;
+        } else {
+            Node<T> x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x.value;
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexWithEquals(index);
+
+        Node<T> current = getElementByIndex(index);
+        Node<T> prev = current.prev;
+        Node<T> next = current.next;
+        Node<T> newNode = new Node<>(prev, value, next);
+        T oldValue = current.value;
+        current = null;
+
+        if (prev == null) {
+            first = newNode;
+        } else if (next == null) {
+            last = newNode;
+        } else {
+            prev.next = newNode;
+            next.prev = newNode;
+        }
+
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexWithEquals(index);
+
+        Node<T> current = getElementByIndex(index);
+        Node<T> next = current.next;
+        Node<T> prev = current.prev;
+
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            current.prev = null;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            current.next = null;
+        }
+        T oldValue = current.value;
+        current = null;
+        size--;
+
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        int k = 0;
+        if (object == null) {
```

Comment:
let's have single for loop here. We just need to improve if check so it can compare both null and non null values

Reply:
Roman, thanks for review!

Reply:
> Roman, thanks for review!

you're welcome 🤗